File: spec/java/src/io/kaitai/struct/specwrite/TestStrEncodings.java
package io.kaitai.struct.specwrite;

import io.kaitai.struct.ConsistencyError;
import io.kaitai.struct.KaitaiStruct.ReadWrite;
import io.kaitai.struct.testwrite.StrEncodings;
import org.testng.annotations.Test;
import static org.testng.Assert.*;

public class TestStrEncodings extends CommonSpec {
    @Test
    public void testCheckNull() throws Exception {
        final StrEncodings r = new StrEncodings();

        r.setStr1("woo");
        r.setLenOf1(3);

        r.setLenOf2(15);

        Throwable thr = expectThrows(NullPointerException.class, new ThrowingRunnable() {
            @Override
            public void run() throws Throwable {
                r._check();
            }
        });
        final String msg = thr.getMessage();
        if (msg != null) {
            final String fieldName = "str2";
            assertTrue(msg.matches(".*\\b" + fieldName + "\\b.*"), "expected the error message to contain '" + fieldName + "', but got [" + msg + "]");
        }
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str2,.*")
    public void testCheckMismatch() throws Exception {
        StrEncodings r = new StrEncodings();

        r.setStr1("Some ASCII");
        r.setStr2("こんにちは");
        r.setStr3("こんにちは");
        r.setStr4("░▒▓");

        // To be auto-derived
        r.setLenOf1(10);
        r.setLenOf2(12); // should be 15
        r.setLenOf3(10);
        r.setLenOf4(3);

        r._check();
    }

    @Override
    protected Class<? extends ReadWrite> getStructClass() {
        return StrEncodings.class;
    }

    @Override
    protected String getSrcFilename() {
        return "str_encodings.bin";
    }
}

File: spec/python/specwrite/test_str_encodings.py
# -*- coding: utf-8 -*-
import unittest
from kaitaistruct import ConsistencyError
from specwrite.common_spec import CommonSpec

from testwrite.str_encodings import StrEncodings

class TestStrEncodings(CommonSpec.Base):
    def __init__(self, *args, **kwargs):
        super(TestStrEncodings, self).__init__(*args, **kwargs)
        self.struct_class = StrEncodings
        self.src_filename = 'src/str_encodings.bin'

    def test_check_null(self):
        r = StrEncodings()

        r.str1 = "woo"
        r.len_of_1 = 3

        r.len_of_2 = 15

        with self.assertRaisesRegex(AttributeError, u"\\bstr2\\b"):
            r._check()

    def test_check_mismatch(self):
        r = StrEncodings()

        r.str1 = u"Some ASCII"
        r.str2 = u"こんにちは"
        r.str3 = u"こんにちは"
        r.str4 = u"░▒▓"

        # To be auto-derived
        r.len_of_1 = 10
        r.len_of_2 = 12  # should be 15
        r.len_of_3 = 10
        r.len_of_4 = 3

        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str2,"):
            r._check()
